utils/src/array.rs

```diff
@@ -0,0 +1,122 @@
+//! This module provides different helpers in creating constant sized
+//! arrays and converting them to different formats.
+
+// @volhovm It could potentially be more efficient with unsafe tricks.
+/// Converts a two-dimensional vector to a constant sized two-dimensional array.
+pub fn vec_to_boxed_array2<T, const N: usize, const M: usize>(
+    vec: Vec<Vec<T>>,
+) -> Box<[[T; N]; M]> {
+    let vec_of_slices2: Vec<[T; N]> = vec
+        .into_iter()
+        .map(|x: Vec<T>| {
+            let y: Box<[T]> = x.into_boxed_slice();
+            let z: Box<[T; N]> = y
+                .try_into()
+                .unwrap_or_else(|_| panic!("vec_to_boxed_array2: length mismatch inner array"));
+            *z
+        })
+        .collect();
+    let array: Box<[[T; N]; M]> = vec_of_slices2
+        .into_boxed_slice()
+        .try_into()
+        .unwrap_or_else(|_| panic!("vec_to_boxed_array2: length mismatch outer array"));
+
+    array
+}
+
+/// Converts a three-dimensional vector to a constant sized two-dimensional array.
+pub fn vec_to_boxed_array3<T, const N: usize, const M: usize, const K: usize>(
+    vec: Vec<Vec<Vec<T>>>,
+) -> Box<[[[T; N]; M]; K]> {
+    let vec_of_slices2: Vec<[[T; N]; M]> =
+        vec.into_iter().map(|v| *vec_to_boxed_array2(v)).collect();
+    vec_of_slices2
+        .into_boxed_slice()
+        .try_into()
+        .unwrap_or_else(|_| panic!("vec_to_boxed_array3: length mismatch outer array"))
+}
+
+/// A macro similar to `vec![$elem; $size]` which returns a boxed
+/// array, allocated directly on the heap (via a vector).
+///
+/// ```rustc
+///     let _: Box<[u8; 1024]> = box_array![0; 1024];
+/// ```
+///
+/// See
+/// <https://stackoverflow.com/questions/25805174/creating-a-fixed-size-array-on-heap-in-rust/68122278#68122278>
+#[macro_export]
+macro_rules! box_array {
+    ($val:expr ; $len:expr) => {{
+        // Use a generic function so that the pointer cast remains type-safe
+        fn vec_to_boxed_array<T>(vec: Vec<T>) -> Box<[T; $len]> {
```

Comment:
As it does use unsafe, I would use `unsafe_` as a prefix in the function name, so the user is aware of this.